boilerplate/openshift/golang-osd-operator/csv-generate/common-generate-operator-bundle.py

```diff
@@ -41,25 +41,159 @@
+if not os.path.exists(VERSION_DIR):
+    os.mkdir(VERSION_DIR)
+
+with open('config/templates/csv-template.yaml', 'r') as stream:
+    csv = yaml.safe_load(stream)
+
+# by_kind is a map, keyed by Kind, of all the yaml documents we find
+# under the deploy/ directory. We need to load them all before we start
+# processing because some (e.g. a ClusterRole and the ServiceAccount in
+# its corresponding ClusterRoleBinding) are interdependent.
+by_kind = {}
+# crb_by_cr is a map, keyed by ClusterRole name, of ClusterRoleBinding
+# documents.
+crb_by_cr = {}
+# rb_by_role is a map, keyed by Role name, of RoleBinding documents
+rb_by_role = {}
+for d, _, files in os.walk('deploy'):
+    for fname in files:
+        if not fname.endswith('.yaml'):
+            continue
+        path = os.path.join(d, fname)
+        with open(path, 'r') as stream:
+            for doc in yaml.safe_load_all(stream):
+                print(
+                    f"Loading {doc['kind']} {doc['metadata']['name']} " +
+                    f"from {path}")
+                if doc['kind'] not in by_kind:
+                    by_kind[doc['kind']] = []
+                by_kind[doc['kind']].append(doc)
+                if doc['kind'] == 'ClusterRoleBinding':
+                    crb_by_cr[doc['roleRef']['name']] = doc
+                if doc['kind'] == 'RoleBinding':
+                    rb_by_role[doc['roleRef']['name']] = doc
+
+class NoServiceAccountSubjectInBinding(Exception):
+    def __init__(self, binding):
+        super().__init__(
+            f"No ServiceAccount Subject in the following Binding:\n{binding}")
+
+class NoDeploymentFound(Exception):
+    def __init__(self):
+        super().__init__("At least one Deployment is required!")
+
+class MultipleDeploymentsNotSupported(Exception):
+    def __init__(self, deployments):
+        super().__init__(
+            f"Multiple Deployments not supported! Found {len(deployments)}.")
+
+class BindingsNotSupported(Exception):
+    def __init__(self, bindings):
+        super().__init__(
+            "[Cluster]RoleBindings are only supported when they correspond " +
+            "to provided [Cluster]Roles. Found the following orphans:\n" +
+            f"{bindings}")
+
+def log_resource(resource):
+    """Log a message that we're processing the given resource.
+
+    :param resource: A k8s resource definition dict, expected to have at least
+        a kind and metadata.name.
+    """
+    print(f"Processing {resource['kind']} {resource['metadata']['name']}")
+
+def filename_for(resource):
+    """Generate a unique base file name for the given resource.
+
+    :param resource: A k8s resource definition dict, expected to have at least
+        a kind and metadata.name.
+    :return: A base file name (no directory) that should be unique for the
+        given resource (assuming the resource is itself unique).
+    """
+    return '_'.join((resource['kind'], resource['metadata']['name'])) + '.yaml'
```

Comment:
Slight concern that this resource may overwrite another if its the same kind+name but different namespace. Include namespace if available or append unique identifier?

Reply:
Or possibly just check if the path already exists in `bundle()` and error.

Reply:
Yes, good call. I actually started doing that this morning when I was working on the ServiceAccount paring thing we decided not to do. I'll resurrect it.